apps/common/lib/lexical/ast/detection.ex

```diff
@@ -65,6 +66,20 @@ defmodule Lexical.Ast.Detection do
     end)
   end
 
+  def ast_to_range(ast) do
+    ast_to_range(ast, 0, 0)
+  end
+
+  def ast_to_range(ast, start_offset, end_offset) do
+    %{start: [line: start_line, column: start_col], end: [line: end_line, column: end_col]} =
+      Sourceror.get_range(ast)
```

Comment:
`Sourceror.get_range/1` can return `nil`, which needs to be handled here. It looks like we already have a `Lexical.Ast.Range.fetch/2`, which I think should be used instead. It could be modified to support the start/end offsets.

This will obviously mean we'll have to modify the usage in the detection modules to handle an `:error` case.

Reply:
yes, but fetch_range takes a document, which makes the api cumbersome.
Can you provide an example of sourceror returning nil in these cases? I know it's possible, but I was under the impression that it will only do so if there is no information in the AST, which is not the case for our usages.

Reply:
> yes, but fetch_range takes a document, which makes the api cumbersome.

Fair!

> Can you provide an example of sourceror returning nil in these cases? 

So the detection abstractions are new to me and it's definitely possible that `nil` doesn't need to be handled in these cases. That said, here's one example and what `get_range` returns:

```elixir
iex(15)> {:ok, ast, _} = Lexical.Ast.from(~S|"foo#{}"|)
{:ok,
 {:<<>>, [delimiter: "\"", line: 1, column: 1],
  [
    "foo",
    {:"::", [line: 1, column: 5],
     [
       {{:., [line: 1, column: 5], [Kernel, :to_string]},
        [closing: [line: 1, column: 7], line: 1, column: 5],
        [{:__block__, [], []}]},
       {:binary, [line: 1, column: 5], nil}
     ]}
  ]}, []}

iex(17)> ast |> Macro.prewalker() |> Enum.map(&{Sourceror.get_range(&1), &1})
[
  {%{end: [line: 1, column: 9], start: [line: 1, column: 1]},
   {:<<>>, [delimiter: "\"", line: 1, column: 1],
    [...snip]}},

  {nil, "foo"},

  {%{end: [line: 1, column: 11], start: [line: 1, column: 1]},
   {:"::", [line: 1, column: 5],
    [...snip]}},

  {%{end: [line: 1, column: 8], start: [line: 1, column: 1]},
   {{:., [line: 1, column: 5], [Kernel, :to_string]},
    [closing: [line: 1, column: 7], line: 1, column: 5],
    [{:__block__, [], []}]}},

  {nil, {:., [line: 1, column: 5], [Kernel, :to_string]}},

  {nil, Kernel},

  {nil, :to_string},

  {nil, {:__block__, [], []}},

  {%{end: [line: 1, column: 11], start: [line: 1, column: 5]},
   {:binary, [line: 1, column: 5], nil}}
]
```

Reply:
Thanks, that's helpful, let me think on this a bit.